### DISCUSSION TITLE
[Bug] Korean Zero-shot training

### DISCUSSION THREAD

**Discussion by hathubkhn** [coqui-ai/TTS#2291] [General Q&A]
### Describe the bug

Dear Author, 
Thank you for your nice work, 
I tried apply CoquiTTS for Korean languages. And I also trained this with KSS single dataset, (12844 samples) and seem that it can w
[coquiTTS.zip](https://github.com/coqui-ai/TTS/files/10173335/coquiTTS.zip)
ork and synthesize with an acceptable result with GlowTTS and MB-MelGan. The audio result is attached
However, when I tried to apply to multi-speaker data (AIHUB) with more thatn 1500 speakers and 30000 samples only, I get the error. Could you help me figure out what is problem. And the better way to do Zero-shot for Korean. The error image will be attached.
[coquiTTS.zip](https://github.com/coqui-ai/TTS/files/10173341/coquiTTS.zip)

### To Reproduce

[src.zip](https://github.com/coqui-ai/TTS/files/10173361/src.zip)

### Environment

```shell
-TTS develop version, using TTS folder, and not install TTS
- cuda117
- python 3.7
```

**Comment by p0p4k** [comment]
Hello, in your config for glowtts or any tts model that you are using, try to set 'num_eval_loader_workers' as 0.
Or find some variable 'num_workers' and set it to 0.

**Comment by hathubkhn** [comment]
Hello, 
Thank you for your reply.
And I also have other questions about the correct way to train multispeaker with coquiTTS for voice cloning purposes.
As my understanding, in glowtts config file, we have some values for setup speaker_embedding, could you help me figure out this?
 When I check out yourTTS colab program, I understand that we dont need to fine-tuning with new voice, we only calculate the speaker embedding of the new voice and synthesize with TTS checkpoint, encoder checkpoint, and vocoder checkpoint. Is it true?
 Thank you very much

**Comment by Edresson** [comment]
> e other questions about the correct way to train multispeaker with coquiTTS for voice cloning purposes.
> As my understanding, in glowtts config file, we have some values for setup speaker_embedding, could you help me figure out this?
> When I check out yourTTS colab program, I understand that we dont need to fine-tuning with new voice, we only calculate the speaker embedding of

Hi @hathubkhn, 
It is right, YourTTS can produce a new voice without the model being trained in the target voice. During the training, the YourTTS model is conditioned with speaker embeddings extracted from a speaker encoder (trained with thousand of speakers). The speaker encoder can generalize good embeddings for new speakers and the YourTTS model which is conditioned with these embeddings can generate new voices as well. If you are interested in understanding how this works @WeberJulian  did a Youtube video about the YourTTS model and you can watch it [here](https://www.youtube.com/watch?v=sCXOUhnfzxM). In addition, have my talk at Nvidia's AI submit that you can access [here](https://www.nvidia.com/en-us/events/speech-ai-summit/).

**Comment by p0p4k** [comment]
@Edresson thanks a lot for summarising the process. I can refer people to this comment from now on.

**Comment by hathubkhn** [comment]
Hello,
Thank you for reply me, 
So, Please review my task when I am trying to apply coquiTTS for Korean voice cloning task. I am worry that I have misunderstood somewhere. 
1.  I turn on "use_speaker_embedding = true" in the model config file

2.  Training model with multispeaker data with dataset format like this:
     wavs/6083/6083_G1A1E7_MHB/6083_G1A1E7_MHB_002157.wav|근방에 맛있는 집 추천해드립니다.|근방에 맛있는 집 추천해드립니다.|6083
wavs/2230/2230_G2A2E7_LJI/2230_G2A2E7_LJI_000590.wav|알겠습니다, 오전 일곱 시에 알람 예약합니다.|알겠습니다, 오전 일곱 시에 알람 예약합니다.|2230
wavs/8553/8553_G1A1E7_HHJ/8553_G1A1E7_HHJ_001090.wav|내일 열시 퇴실입니다.|내일 열시 퇴실입니다.|8553

3. And in synthesize part, I just call this, with model is glowTTS, vocoder is MB-MelGan, Speaker Encoder I download from other source(however it trained with English dataset)

synthesizer = Synthesizer(
    MODEL_PATH,
    CONFIG_PATH,
    None,
    None,
    vocoder_path,
    vocoder_config,
    SE_MODEL_PATH,
    SE_CONFIG_PATH,
    False
)
![Screenshot 2022-12-09 at 1 04 53 PM](https://user-images.githubusercontent.com/30932833/206621619-7a1fd570-894f-465b-97be-9449b391d9e7.png)


wav = synthesizer.tts(text,  None, speaker_wav = ['resample/data/wavs/tts1_01.wav']) 



=> However, I have a problem when synthesize, the image is attached. Could you help me check it, I dont know why?

**Comment by p0p4k** [comment]
So, I think you have got is confused a little bit. 

Speaker embedding means, inside the TTS model, there will be a layer at the start (just like a text encoder), that converts the speaker index into an embedding (trying to capture speaker style, etc.).
In other case, if you use an external SE (speaker encoder), it will do what the embedding layer did, except it is pre-trained on another dataset (i believe it is celeb-voice dataset or something like that). The TTS model then receives that output from SE and combines with the text embedding/encoding and then proceeds normally.

If you are using SE_MODEL_PATH, SE_CONFIG_PATH -> (SE == Speaker Encoder), then no need to use "speaker embedding".
But if you trained model with a speaker embedding, then no need to provide SE_MODEL_PATH and SE_CONFIG_PATH.

I hope my understanding is correct and not misleading you. Hopefully, some higher ups can add to this comment, if needed. 
(PS. try to post the error logs/ runs by copy pasting text from the console, since text is lighter than a picture and takes less server space, plus it is easier for other to search related line for debugging. Thanks.)

**Comment by hathubkhn** [comment]
Hello 

> So, I think you have got is confused a little bit.
> 
> Speaker embedding means, inside the TTS model, there will be a layer at the start (just like a text encoder), that converts the speaker index into an embedding (trying to capture speaker style, etc.). In other case, if you use an external SE (speaker encoder), it will do what the embedding layer did, except it is pre-trained on another dataset (i believe it is celeb-voice dataset or something like that). The TTS model then receives that output from SE and combines with the text embedding/encoding and then proceeds normally.
> 
> If you are using SE_MODEL_PATH, SE_CONFIG_PATH -> (SE == Speaker Encoder), then no need to use "speaker embedding". But if you trained model with a speaker embedding, then no need to provide SE_MODEL_PATH and SE_CONFIG_PATH.
> 
> I hope my understanding is correct and not misleading you. Hopefully, some higher ups can add to this comment, if needed. (PS. try to post the error logs/ runs by copy pasting text from the console, since text is lighter than a picture and takes less server space, plus it is easier for other to search related line for debugging. Thanks.)
[log_infer.log](https://github.com/coqui-ai/TTS/files/10192770/log_infer.log)

Hello [p0p4k](https://github.com/p0p4k), 
1.Yes, I think you are correct. And I have other question like should we use the existing SE model or use speaker_embedding from scratch? 

2. I attached my log, and please give me some recommendations, thank you very much!

**Comment by p0p4k** [comment]
Hello,

> 1.Yes, I think you are correct. And I have other question like should we use the existing SE model or use speaker_embedding from scratch?

It depends, try out both and judge the results yourself. Speaker embedding will just learn the speakers from your dataset, but the SE-external can generalize to any voice, I think.

> 2. I attached my log, and please give me some recommendations, thank you very much!

If you trained with SE-external, use same config. If trained with speaker embedding, use that config. Other than that, using default values will still give good results. 
I had a plan to train Korean data from AIHUB earlier this year, but got busy with other stuff. If you train this model and do not mind sharing it, we can try to improve upon it later on.

Good luck.

**Comment by hathubkhn** [comment]
Hello, 
1. So, if I use the SE model, I need to calculate d_vector_file, right?
2. Do you know what is problem in my program for now?
Sure, I need other professional person to help me finish this task perfectly, :(( Could I contact with you directly?

**Comment by p0p4k** [comment]
Add me on Gitter (you can use Gitter in browser directly) 
(username : p0p4k)

**Comment by hathubkhn** [comment]
Hello, 
I alreay added and sent message in Gitter. Could we help some discussions?
Thank you very much

**Comment by Edresson** [comment]
@hathubkhn Recently, we created a recipe that makes everything easier. If you like you can try to fine-tune the model with this recipe.

The recipe replicates the first experiment proposed in the YourTTS paper. The recipe replicates the single language training using the VCTK dataset (it downloads, resamples, and extracts the speaker embeddings automatically :)). However, if you are interested in multilingual training, we have commented on parameters on the VitsArgs class instance that should be enabled for multilingual training: https://github.com/coqui-ai/TTS/blob/dev/recipes/vctk/yourtts/train_yourtts.py

**Comment by hathubkhn** [comment]
> @hathubkhn Recently, we created a recipe that makes everything easier. If you like you can try to fine-tune the model with this recipe.
> 
> The recipe replicates the first experiment proposed in the YourTTS paper. The recipe replicates the single language training using the VCTK dataset (it downloads, resamples, and extracts the speaker embeddings automatically :)). However, if you are interested in multilingual training, we have commented on parameters on the VitsArgs class instance that should be enabled for multilingual training: https://github.com/coqui-ai/TTS/blob/dev/recipes/vctk/yourtts/train_yourtts.py

Hello Edresson, 
I am doing on Korean languages, so could I use the way like:
1. Training VITS with training data of one single woman speaker (~3 hours long) => save checkpoint
2. Create a model with saved checkpoint, and re-trained with one men speaker (~17min long)

**Comment by Edresson** [comment]
> 

Why not add the samples from both speakers in the same training? YourTTS is a multi-speaker model it can deal with differents voices without problems.

**Reply by hathubkhn** [reply]
> > 
> 
> Why not add the samples from both speakers in the same training? YourTTS is a multi-speaker model it can deal with differents voices without problems.

Actually, my final goal is cloning voice. For now, I am choosing the data like that:
- 100 woman speakers and 100 men speakers
- Each speaker has 1 hour of long audio
- Total: 213000 training samples (~200 hours)
-> Then, I used compute_embedding.py to calculate d_vector_files.json file before training. However, I see the performance is not good. Could you help me check the reason? And do you know any denoise processing for process data before training. 

Thank you very much,
